src/tool/core/generate.rs

```diff
@@ -0,0 +1,130 @@
+            info!("Using '{specific_problem}' as a slug");
+            Cow::Borrowed(specific_problem)
+        }
+    } else {
+        // Daily problem
+        let slug = daily_challenge::get_daily_challenge_slug()?;
+        info!("Slug for daily problem is: '{slug}'");
+        Cow::Owned(slug)
+    };
+
+    let (module_name, module_code) = create_module_code(title_slug, args)
+        .context("Failed to generate the name and module code")?;
+    write_to_disk::write_file(&module_name, module_code).context("Failed to write to disk")?;
+    println!("Generated module: {module_name}");
+    Ok(())
+}
+
+/// Gets the code and other data from leetcode and generates the suitable code for the module and the name of the module
+/// Returns the module name and the module code
+///
+/// NB: Did not return `Cow` because `module_name` is always a modified version of the input
+pub fn create_module_code(
+    title_slug: Cow<String>,
+    args: &cli::GenerateArgs,
+) -> anyhow::Result<(String, String)> {
+    info!("Building module contents for {title_slug}");
+
+    let meta_data =
+        get_problem_metadata(&title_slug).context("Failed to retrieve problem meta data")?;
+
+    // Add problem URL
+    let mut code_snippet = format!(
+        "//! Solution for {}{title_slug}\n",
+        Config::LEETCODE_PROBLEM_URL
+    );
+
+    // Add problem number and title
+    code_snippet.push_str(&format!(
+        "//! {}\n",
+        meta_data
+            .get_num_and_title()
+            .context("Failed to get problem number and title")?
+    ));
+
+    // Add blank line between docstring and code
+    code_snippet.push('\n');
+
+    // Get code snippet
+    let problem_code = get_code_snippet_for_problem(&title_slug)?;
+    code_snippet.push_str(problem_code.as_ref());
+
+    // Add 2 empty lines between code and "other stuff (like tests and struct definition"
+    code_snippet.push_str("\n\n");
+
+    // Add struct for non design questions
+    if problem_code.type_.is_non_design() {
+        code_snippet.push_str("\nstruct Solution;\n")
+    }
+
+    // Add leet code types
+    if problem_code.has_tree() {
+        code_snippet.push_str("use cargo_leet::TreeNode;\n");
+    }
+    if problem_code.has_list() {
+        code_snippet.push_str("use cargo_leet::ListNode;\n");
+    }
+
+    // Add tests
+    let tests = meta_data.get_test_cases(&problem_code)?;
+    code_snippet.push_str(&tests);
+
+    // Set module name
+    let module_name = if args.should_include_problem_number {
+        info!("Including problem number in module name");
+        format!(
+            "_{}_{}",
+            meta_data.get_id()?,
+            title_slug.to_case(Case::Snake)
+        )
+    } else {
+        info!("Using snake case slug for module name");
+        title_slug.to_case(Case::Snake)
+    };
+
+    Ok((module_name, code_snippet))
+}
+
+/// Quick and dirty test to see if this is a url
+fn is_url(value: &str) -> bool {
+    value.contains('/')
```

Comment:
why not starts with `https://`

Reply:
Yeah we can chance this. I didn't do https in case it was http. I just picked a character that wasn't allowed as part of the slug and must be in the url and used that the differentiate.  I check for the correct URL prefix after.

Reply:
As per our discussion I added the top level context (required me to move the code into a function to keep it easy to read)